subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -484,36 +482,98 @@ public TaskInfo getTaskToExecute() {
+                final AtomicBoolean restartIteration = new AtomicBoolean();
+                final Set<String> lockedProjects = Sets.newHashSet();
                 while (iterator.hasNext()) {
-                    TaskInfo taskInfo = iterator.next();
-                    allTasksComplete = allTasksComplete && taskInfo.isComplete();
-                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete() && canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
-                        nextMatching = taskInfo;
-                        iterator.remove();
-                        break;
+                    final TaskInfo taskInfo = iterator.next();
+                    Project project = taskInfo.getTask().getProject();
+                    final String projectPath = ((ProjectInternal) project).getIdentityPath().toString();
+                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete()) {
+                        try {
+                            final ProjectLock projectLock = projectLockService.getProjectLock(projectPath);
+                            boolean couldNotLock = !projectLock.tryWithProjectLock(new Runnable() {
+                                @Override
+                                public void run() {
+                                    boolean canExecute = false;
+                                    try {
+                                        if (lockedProjects.contains(projectLock.getProjectPath())) {
+                                            // If the state of project locks has changed, we want to restart
+                                            // task selection so that we can be sure that the earliest task
+                                            // from this project in the execution queue is selected first.
+                                            restartIteration.set(true);
+                                        } else {
+                                            if (canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
+                                                selected.set(true);
+                                                iterator.remove();
+                                                if (taskInfo.allDependenciesSuccessful()) {
+                                                    taskInfo.startExecution();
+                                                    recordTaskStarted(taskInfo);
+                                                    canExecute = true;
+                                                } else {
+                                                    taskInfo.skipExecution();
+                                                }
+                                            }
+                                        }
+                                    } finally {
+                                        // We need to unlock the plan before unlocking the project.  This prevents a deadlock
```

Comment:
This is where we want to end up, not necessarily where we want to start. We can simplify things to look something like:

- when a task is available, and a cpu resource lock is available, and the relevant project lock is available, atomically mark them as in use by the worker:
- worker starts running the task
- worker decides the actions don't need to be run, and:
    - prepares the outputs, by fetching them from the build cache (still holding the resource locks).
- else, the worker runs the actions of the task:
    - worker finishes running the actions, there may be worker actions running.
    - worker waits for all worker actions to complete (releasing the project lock and the cpu resource lock but not the task state)
    - worker finishes up, snapshotting outputs (and still holding the project lock and cpu lock)
- task is marked as complete and worker gives up its project lock and cpu lock and task state.

So, if it's simpler, we can hold all of the resource locks across the whole execution if we want to, releasing them all while waiting for the child worker actions to complete and reacquiring them all to finish up.